squidpy/pl/_spatial.py

```diff
@@ -0,0 +1,764 @@
+
+
+def _get_spatial_attrs(
+    adata: AnnData,
+    spatial_key: str = Key.obsm.spatial,
+    library_id: Optional[Sequence[str] | None] = None,
+    img: Optional[Sequence[NDArrayA] | NDArrayA | None] = None,
+    img_key: str | None = None,
+    scale_factor: Optional[Sequence[float] | float | None] = None,
+    size: Optional[Sequence[float] | float | None] = None,
+    size_key: str = Key.uns.size_key,
+    bw: bool = False,
+) -> Tuple[Sequence[str], Sequence[float], Sequence[float], Sequence[NDArrayA]]:
+    """Return lists of image attributes saved in adata for plotting."""
+    library_id = Key.uns.library_id(adata, spatial_key, library_id, return_all=True)
+    if library_id is None:
+        raise ValueError(f"Could not fetch `library_id`, check that `spatial_key: {spatial_key}` is correct.")
+
+    image_mapping = Key.uns.library_mapping(adata, spatial_key, Key.uns.image_key, library_id)
+    scalefactor_mapping = Key.uns.library_mapping(adata, spatial_key, Key.uns.scalefactor_key, library_id)
+    scalefactors = _get_unique_map(scalefactor_mapping)
+
+    if not (image_mapping.keys() == scalefactor_mapping.keys()):  # check that keys match
+        raise KeyError(
+            f"Image keys: `{image_mapping.keys()}` and scalefactor keys: `{scalefactor_mapping.keys()}` are not equal."
+        )
+
+    if img_key is None:
+        img_key = _get_unique_map(image_mapping)  # get intersection of image_mapping.values()
+        img_key = img_key[0]  # get first of set
+    else:
+        if img_key not in image_mapping.values():
+            raise ValueError(f"Image key: `{img_key}` does not exist. Available image keys: `{image_mapping.values()}`")
+
+    if img is None:
+        img = [adata.uns[Key.uns.spatial][i][Key.uns.image_key][img_key] for i in library_id]
+    else:  # handle case where img is ndarray or list
+        img = _maybe_get_list(img, np.ndarray, library_id)
+
+    if bw:
+        img = [np.dot(im[..., :3], [0.2989, 0.5870, 0.1140]) for im in img]
+
+    if scale_factor is None:  # get intersection of scale_factor and match to img_key
+        scale_factor_key = [i for i in scalefactors if img_key in i]
+        if len(scale_factor_key) == 0:
+            raise ValueError(f"No `scale_factor` found that could match `img_key`: {img_key}.")
+        _scale_factor_key = scale_factor_key[0]  # get first scale_factor
+        scale_factor = [adata.uns[Key.uns.spatial][i][Key.uns.scalefactor_key][_scale_factor_key] for i in library_id]
+    else:  # handle case where scale_factor is float or list
+        scale_factor = _maybe_get_list(scale_factor, float, library_id)
+
+    # size_key = [i for i in scalefactors if size_key in i][0]
+    if size_key not in scalefactors and size is None:
+        raise ValueError(
+            f"Specified `size_key: {size_key}` does not exist and size is `None`,\
+            available keys are: `{scalefactors}`.\n Specify valid `size_key` or `size`."
+        )
+    if size is None:
```

Comment:
Why can size be `None`, just make it default `1.0`

Reply:
because in the case of non-visium data, if None it computes optimal one based on scanpy

```size = 120000 / adata.shape[0] if size is None else size```

Reply:
Ok, sounds good.